File: coffee-system/src/main/java/com/jesperancinha/coffee/system/manager/PreProcessorImpl.java
package com.jesperancinha.coffee.system.manager;

import com.jesperancinha.coffee.api.concurrency.PreActionCallable;
import com.jesperancinha.coffee.api.concurrency.StartupCallable;
import com.jesperancinha.coffee.api.manager.MachineProcessor;
import com.jesperancinha.coffee.api.manager.PreProcessor;
import com.jesperancinha.coffee.system.concurrency.ActionCallable;
import com.jesperancinha.coffee.system.concurrency.PreActionCallableImpl;
import com.jesperancinha.coffee.system.concurrency.StartupCallableImpl;
import com.jesperancinha.coffee.system.input.CoffeeMachines.CoffeMachine.Coffees.Coffee;
import com.jesperancinha.coffee.system.input.CoffeeMachines.CoffeMachine.PaymentTypes.Payment;
import com.jesperancinha.coffee.system.input.Employees.Employee;
import com.jesperancinha.coffee.system.input.Employees.Employee.Actions.PostAction;
import com.jesperancinha.coffee.system.input.Employees.Employee.Actions.PreAction;
import com.jesperancinha.coffee.system.queues.Queue;
import com.jesperancinha.coffee.system.queues.QueuePreActivityImpl;
import lombok.Getter;
import lombok.experimental.Accessors;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.concurrent.Callable;

@Accessors(chain = true)
@Getter
@Service
public class PreProcessorImpl extends ProcessorAbstract implements PreProcessor {

    private final StartupCallable startupCallable;
    @Autowired
    private QueuePreActivityImpl queuePreActivity;
    @Autowired
    private MachineProcessor machineProcessor;

    PreProcessorImpl() {
        super();
        startupCallable = new StartupCallableImpl();
    }

    @Override
    public void callPreActions(
            final Employee employee,
            final String name,
            final List<PreAction> actions,
            final Coffee coffee, Payment payment,
            final List<PostAction> postActions
    ) {

        final PreActionCallable preActionCallable = new PreActionCallableImpl(
                employee,
                name,
                coffee,
                payment,
                postActions,
                machineProcessor
        );
        actions.forEach(
                preActionCallable::addPreAction
        );
        startupCallable.getAllCallables().add(preActionCallable);
    }

    @Override
    public void runAllCalls() {
        runAllCalls(startupCallable);
    }

    @Override
    public void waitForAllCalls() {
        waitForAllCalls(startupCallable);
    }

    @Override
    public Queue getExecutorServiceQueue() {
        return queuePreActivity;
    }

    @Override
    public String getExecutorName(Callable<Boolean> callable) {
        return ((ActionCallable) callable).getName();
    }

    @Override
    public void addQueueSize(int queueSize, String name) {
        queuePreActivity.setQueueSize(queueSize, name);
    }

    @Override
    public void initExecutors() {
        queuePreActivity.initExecutors();
    }

    @Override
    public void stopExectutors() {
        queuePreActivity.stopExecutors();
    }
}

File: coffee-system/src/main/kotlin/org/jesperancinha/coffee/system/manager/PreProcessorImpl.kt
package org.jesperancinha.coffee.system.manager

import lombok.Getter
import lombok.experimental.Accessors
import org.jesperancinha.coffee.system.api.concurrency.QueueCallable
import org.jesperancinha.coffee.system.concurrency.ActionCallable
import org.jesperancinha.coffee.system.concurrency.PreActionCallableImpl
import org.jesperancinha.coffee.system.concurrency.StartupCallableImpl
import org.jesperancinha.coffee.system.input.CoffeeMachines.CoffeMachine.Coffees.Coffee
import org.jesperancinha.coffee.system.input.CoffeeMachines.CoffeMachine.PaymentTypes.Payment
import org.jesperancinha.coffee.system.input.Employees.Employee
import org.jesperancinha.coffee.system.input.Employees.Employee.Actions.PostAction
import org.jesperancinha.coffee.system.queues.Queue
import org.jesperancinha.coffee.system.queues.QueuePreActivityImpl
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.concurrent.Callable
import java.util.function.Consumer

@Accessors(chain = true)
@Getter
@Service
abstract class PreProcessorImpl internal constructor() : ProcessorAbstract() {
    private val startupCallable: StartupCallableImpl

    @Autowired
    private val queuePreActivity: QueuePreActivityImpl? = null

    @Autowired
    private val machineProcessor: MachineProcessorImpl? = null

    init {
        startupCallable = StartupCallableImpl()
    }

    fun callPreActions(
        employee: Employee,
        name: String?,
        actions: List<org.jesperancinha.coffee.system.input.Employees.Employee.Actions.PreAction?>,
        coffee: Coffee, payment: Payment,
        postActions: List<PostAction?>
    ) {
        val preActionCallable = PreActionCallableImpl(
            employee,
            name,
            coffee,
            payment,
            postActions,
            machineProcessor
        )
        actions.forEach(Consumer<org.jesperancinha.coffee.system.input.Employees.Employee.Actions.PreAction> { preAction: org.jesperancinha.coffee.system.input.Employees.Employee.Actions.PreAction ->
            preActionCallable.addPreAction(
                preAction
            )
        })
        startupCallable.allCallables.add(preActionCallable)
    }

    fun runAllCalls() {
        runAllCalls(startupCallable)
    }

    fun waitForAllCalls() {
        waitForAllCalls(startupCallable)
    }

    override val executorServiceQueue: Queue?
        get() = queuePreActivity

    override fun getExecutorName(callable: Callable<Boolean?>): String {
        return (callable as ActionCallable).name
    }

    fun addQueueSize(queueSize: Int, name: String?) {
        queuePreActivity!!.setQueueSize(queueSize, name!!)
    }

    fun initExecutors() {
        queuePreActivity!!.initExecutors()
    }

    fun stopExectutors() {
        queuePreActivity!!.stopExecutors()
    }

    abstract override fun waitForAllCalls(queueCallable: QueueCallable)
    abstract override fun runAllCalls(queueCallable: QueueCallable)
}